proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,221 @@
+used, but the compiler doesn't communicate that to the user.
+It would be great for any HLSL developer to immediately have answers to 
+the following questions:
+For any valid resource type that needs to be bound, what are the set of
+valid register types for that resource, and what are all the possible 
+diagnostics that can be emitted if the resource is given a register 
+type that isn't in the set of valid register types?
+The design below aims to specify the answer to these questions.
+
+## Proposed solution
+Firstly, the most common case is when a resource type that is not a 
+user-defined type is bound as a resource. Any resource will have an 
+associated ResourceAttr attribute, from which we can determine the 
+resource class and resource kind. These two inputs are sufficient to
+determine the expected register type, and any other register type
+will result in an error. The recommended register type will be
+suggested in the diagnostic. The table below specifies what register 
+type will be expected for any resource type that falls under the
+specified resource class. 
+
+| Resource Class | Register Type | Diagnostic |
+|-|-|-|
+| Sampler | s | "Object type '%0' with resource class "Sampler" expects register type 's', but register type '%1' was given." |
+| SRV | t | "Object type '%0' with resource class "SRV" expects register type 't', but register type '%1' was given." |
+| UAV | u | "Object type '%0' with resource class "UAV" expects register type 'u', but register type '%1' was given." |
+| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects register type 'b', but register type '%1' was given." |
+
+There are certain resource kinds within the SRV resource
+class that are "sampleable". These resources are typically
+prefixed with "Texture" in actual HLSL source. So, there will
+be a special check to see whether the candidate resource is 
+prefixed with "Texture", and if so, the "s" register type will
+also be permitted for that resource.
+
+If the given candidate resource is a user-defined type (UDT), then further
+analysis is necessary. The first step is to gather all registers that
+are being bound to this declaration, and collect the register types
+that are being specified. The UDT must have at least one valid resource
+that can be bound to the provided register type. If not, an error must be
+emitted stating that "No resource contained in struct '%0' can be bound
+to register type '%1'". There are no issues if a UDT has more resources
+than there are register binding statements, the resources will be bound to
+the next available space automatically, and so compilation can succeed.
+Additionally, a singular resource can be bound multiple times to different
+spaces, so multiple register types are allowed for a single resource.
+Below is a table that represents the different behaviors that could arise
+given some examples of different UDT's:
+
+| Code | Diagnostic |

```

Comment:
github doesn't seem to be formatting this as a table any more.

I think that this would be much easier to write and read if it were just flat.  Something like:

```
struct Eg1 {
  float f;
  Buffer<float> Buf;
  RWBuffer<float> RWBuf;
};
Eg1 e1 : register(t0) : register(s0); 
// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to s0
```
...
```
struct Eg5 {
  float f;
};

Eg5 e5 : register(t0);
// DefautError warning. "UDF resource 'Eg5' does not contain an applicable resource type for register type 't'.
```

Reply:
Does it look like this for you?

![image](https://github.com/microsoft/hlsl-specs/assets/8118402/b04ad489-2bff-430e-a0a2-35da97600da7)

Note that all the indentation is broken, and the code is all scrunched up on the left so that it wraps and makes it hard to read.